q2_quality_control/quality_control.py

```diff
@@ -122,3 +128,115 @@ def evaluate_taxonomy(output_dir: str, expected_taxa: pd.DataFrame,
                misclassifications=None, underclassifications=None,
                composition_regression=None, score_plot=score_plot,
                mismatch_histogram=None, alignments=None)
+
+
+# decontam added code
+_WHOLE_NUM = (lambda x: x >= 0, 'non-negative')
+_PER_NUM = (lambda x: 1 >= x >= 0, 'between 0 and 1')
+_COL_STR = (lambda x: x in {'column_name', 'column_number'},
+            'sample_name or column_name or column_number')
+_DECON_METHOD_STR = (lambda x: x in {'frequency', 'prevalence', 'combined'},
+                     'freqeuncy, prevalence, combined')
+_BOOLEAN = (lambda x: type(x) is bool, 'True or False')
+# Better to choose to skip, than to implicitly ignore things that KeyError
+_SKIP = (lambda x: True, '')
+_valid_inputs = {
+    'asv_or_otu_table': _SKIP,
+    'meta_data': _SKIP,
+    'threshold': _PER_NUM,
+    'decon_method': _DECON_METHOD_STR,
+    'freq_concentration_column': _SKIP,
+    'prev_control_or_exp_sample_column': _SKIP,
+    'prev_control_sample_indicator': _SKIP,
+}
+
+
+def _check_inputs(**kwargs):
+    for param, arg in kwargs.items():
+        check_is_valid, explanation = _valid_inputs[param]
+        if not check_is_valid(arg):
+            raise ValueError('Argument to %r was %r, should be %s.'
+                             % (param, arg, explanation))
+
+
+def _decontam_identify_helper(track_fp, decon_method):
+    df = pd.read_csv(track_fp, sep='\t', index_col=0)
+    df.index.name = '#OTU ID'
+    # removes last column containing true/false information from the dataframe
+    df = df.drop(df.columns[[(len(df.columns)-1)]], axis=1)
+    if decon_method == 'combined':
+        df = df.fillna(0)
+
+    # removes all columns that are completely empty
+    temp_transposed_table = df.transpose()
+    temp_transposed_table = temp_transposed_table.dropna()
+    df = temp_transposed_table.transpose()
+    metadata = transform(df, from_type=pd.DataFrame,
+                         to_type=DecontamScoreFormat)
+
+    return metadata
+
+
+def decontam_identify(asv_or_otu_table: pd.DataFrame,
+                      meta_data: qiime2.Metadata,
+                      decon_method: str = 'prevalence',
+                      freq_concentration_column: str = 'NULL',
+                      prev_control_or_exp_sample_column: str = 'NULL',
+                      prev_control_sample_indicator: str = 'NULL'
+                      ) -> (DecontamScoreFormat):
+    _check_inputs(**locals())
+    with tempfile.TemporaryDirectory() as temp_dir_name:
+        track_fp = os.path.join(temp_dir_name, 'track.tsv')
+        ASV_dest = os.path.join(temp_dir_name, 'temp_ASV_table.csv')
+        transposed_table = asv_or_otu_table.transpose()
```

Comment:
I think this is one too many transposes, as I'm seeing sample IDs instead of feature IDs in the output when I test locally:

Also these IDs end up being prefixed with an "X"

```tsv
#OTU ID	freq	prev
X00b62637	0.001509783448509	241
X00ce9aa6	0.001903379877358	322
X017d5b5e	0.000885124692441	187
X01a38419	0.000811857689617	155
X01e3a7b5	0.001208878948036	246
...
RB.9.28.22	0.000184459414101	14
RB.9.29.22	0.001260904455254	38
RB.9.8.22.1	0	0
RB.9.8.22.2	1.36418085490761E-05	15
RB.Bucket11.12	0.000441333697915	60
RB.razor.blade	4.36681084938091E-07	2
RB.wipe.tweezer.spatula	2.46721581483982E-06	5
RB5.12.22.1	7.12910814857061E-06	1
```

Reply:
(Or perhaps the R script isn't handling the output correctly and we're just seeing a per-sample summary instead of per-feature?)

Reply:
Oh my yes I realized that the test data that I was working on somehow got transposed before making its way into the 'expected data folder' things should be working now, I corrected the issue in the latest push  https://github.com/qiime2/q2-quality-control/pull/74/commits/bce4616be7568d1aa4cede540b5c0f490ed877f6

Reply:
That explains it! Old-school OTU tables tended to be feature x sample (the BIOM format even does this) instead of sample x feature as it was more practical to "add rows" when manually curating these things. It's an easy thing to get backwards.

Re: checking, there's not technically any way to know for sure, but it generally isn't a problem, as very few people transpose the table, so the orientation is essentially fixed within a standard q2 workflow. (And provenance will reveal the error if someone were to transpose)